plugins/kinetic-dp/fu-kinetic-dp-device.c

```diff
@@ -0,0 +1,283 @@
+/*
+ * Copyright (C) 2015 Richard Hughes <richard@hughsie.com>
+ * Copyright (C) 2016 Mario Limonciello <mario.limonciello@dell.com>
+ * Copyright (C) 2021 Jeffrey Lin <jlin@kinet-ic.com>
+ * Copyright (C) 2022 Hai Su <hsu@kinet-ic.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fcntl.h>
+
+#include "fu-kinetic-dp-common.h"
+#include "fu-kinetic-dp-connection.h"
+#include "fu-kinetic-dp-device.h"
+#include "fu-kinetic-dp-firmware.h"
+#include "fu-kinetic-dp-puma-aux-isp.h"
+#include "fu-kinetic-dp-secure-aux-isp.h"
+
+struct _FuKineticDpDevice {
+	FuUdevDevice parent_instance;
+	FuKineticDpAuxIsp *aux_isp_ctrl;
+	gchar *system_type;
+	FuKineticDpFamily family;
+	FuKineticDpMode mode;
+};
+
+G_DEFINE_TYPE(FuKineticDpDevice, fu_kinetic_dp_device, FU_TYPE_UDEV_DEVICE)
+
+FuKineticDpAuxIsp *
+fu_kinetic_dp_device_get_aux_isp_ctrl(FuKineticDpDevice *self)
+{
+	g_debug("device get aux isp ctrl.");
+	g_return_val_if_fail(FU_IS_DEVICE(self), NULL);
+	return self->aux_isp_ctrl;
+}
+
+/* hookup isp protocol for different chip */
+void
+fu_kinetic_dp_device_set_aux_isp_ctrl(FuKineticDpDevice *self, KtChipId chip_id)
+{
+	g_return_if_fail(FU_IS_KINETIC_DP_DEVICE(self));
+	g_return_if_fail(self->aux_isp_ctrl == NULL);
+
+	if (chip_id == KT_CHIP_JAGUAR_5000 || chip_id == KT_CHIP_MUSTANG_5200) {
+		g_set_object(&self->aux_isp_ctrl,
+			     FU_KINETIC_DP_AUX_ISP(fu_kinetic_dp_secure_aux_isp_new()));
```

Comment:
i don't think the refcounting works here, e.g. i think you need a `g_autoptr(FuDevice) ctrl = fu_kinetic_dp_secure_aux_isp_new()` then you can use `g_set_object(&self->aux_isp_ctrl, ctrl)`.

Reply:
g_autoptr(FuDevice) ctrl will not work because fu_kinetic_dp_secure_aux_isp_new() does not return a FuDevice.

Reply:
`g_autoptr(FuKineticDpSecureAuxIsp) ctrl` then?

Reply:
I tried that and I got a few warning like these.
../plugins/kinetic-dp/fu-kinetic-dp-device.c: In function ‘fu_kinetic_dp_device_set_aux_isp_ctrl’:
/usr/include/glib-2.0/gobject/gobject.h:751:31: warning: assignment to ‘FuKineticDpAuxIsp *’ {aka ‘struct _FuKineticDpAuxIsp *’} from incompatible pointer type ‘FuKineticDpSecureAuxIsp_autoptr’ {aka ‘struct _FuKineticDpSecureAuxIsp *’} [-Wincompatible-pointer-types]
  751 |     (void) (0 ? *(object_ptr) = (new_object), FALSE : FALSE); \

Reply:
I am not sure why it complains about FuKineticDpAuxIsp when I use g_autoptr(FuKineticDpSecureAuxIsp) ctr. It seems to think that I am doing g_autoptr(FuKineticDpAuxIsp) ctrl

Reply:
`g_autoptr(FuKineticDpAuxIsp) ctrl then?`

Reply:
tried that and got a different warning.
[467/850] Compiling C object plugins/k...netic_dp.so.p/fu-kinetic-dp-device.c.o
../plugins/kinetic-dp/fu-kinetic-dp-device.c: In function ‘fu_kinetic_dp_device_set_aux_isp_ctrl’:
../plugins/kinetic-dp/fu-kinetic-dp-device.c:46:44: warning: initialization of ‘FuKineticDpAuxIsp_autoptr’ {aka ‘struct _FuKineticDpAuxIsp *’} from incompatible pointer type ‘FuKineticDpSecureAuxIsp *’ {aka ‘struct _FuKineticDpSecureAuxIsp *’} [-Wincompatible-pointer-types]
   46 |   g_autoptr(FuKineticDpAuxIsp)secure_ctrl =fu_kinetic_dp_secure_aux_isp_new();
      |                                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~